### DISCUSSION TITLE
How to use the agent with a timeout ?

### DISCUSSION THREAD

**Discussion by luc-leclerc-pg** [async-profiler/async-profiler#946] [Q&A] [answered]
#### Context
I am investigating a slow boot time for a java app that only happens in a docker container.

#### Goal
I want to start the agent and have it stop after 60 seconds, to generate the graph. 

#### What I tried
- I found this section on the readme ([launching-as-an-agent](https://github.com/async-profiler/async-profiler?tab=readme-ov-file#launching-as-an-agent)), but it's still not clear how to use it after reading it...
- I setup extra JAVA_OPS to use the agent `JAVA_OPTS="-agentpath:/opt/web-service/async-profiler-3.0-linux-arm64/lib/libasyncProfiler.so=event=cpu,log=log.log,file=profile.html $JAVA_OPTS -d 60"`, but this doesn't work, and the agent doesn't build the profile.html file after 60 seconds.
```
  echo 1 > /proc/sys/kernel/perf_event_paranoid # Allow kernel and user-space measurements. warn: this failed without docker privileged mode
  echo 0 > /proc/sys/kernel/kptr_restrict # Remove restrictions on exposing kernel addresses:
  yum install -y tar gzip
  curl -L https://github.com/async-profiler/async-profiler/releases/download/v3.0/async-profiler-3.0-linux-arm64.tar.gz -o - | tar -xz
  JAVA_OPTS="-agentpath:/opt/web-service/async-profiler-3.0-linux-arm64/lib/libasyncProfiler.so=event=cpu,log=log.log,file=profile.html -XX:+UnlockDiagnosticVMOptions -XX:+DebugNonSafepoints $JAVA_OPTS -d 60"
```

I must be missing really small and obvious but I am not seen it 😢 here is the result. Without the -d 60, I am otherwise able to get it to connect. 
```host-gateway  | + exec java -Xms1024m -Xmx3g -Xss4m -Dsun.net.http.allowRestrictedHeaders=true -Djdk.xml.entityExpansionLimit=0 -agentpath:/opt/web-service/async-profiler-3.0-linux-arm64/lib/libasyncProfiler.so=event=cpu,log=log.log,file=profile.html -XX:+UnlockDiagnosticVMOptions -XX:+DebugNonSafepoints -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/tmp -DautoListenerStartup=false -Dplusgrade.databaseMaintenanceSchedulerEnabled=false -Dlog.sampling.enabled=false -Duser.timezone=UTC -agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=5005 -d 60 -jar host-gateway.jar
host-gateway  | Unrecognized option: -d
host-gateway  | Error: Could not create the Java Virtual Machine.
host-gateway  | Error: A fatal exception has occurred. Program will exit.
host-gateway exited with code 1```

**Comment by apangin** [comment, marked as answer]
`-d` is an option of `asprof`, but not of the agent. In the agent configuration, you may use [`timeout`](https://github.com/async-profiler/async-profiler/blob/v3.0/src/arguments.cpp#L73) argument instead:
```
-agentpath:.../libasyncProfiler.so=event=cpu,file=profile.html,timeout=60s
```
